### ISSUE TITLE
Remove dependency on `hostNetwork` for Alluxio worker Pods

### ISSUE THREAD

**Issue by jiacheliu3** [Alluxio/alluxio#11172]
**Is your feature request related to a problem? Please describe.**
In K8s, if not using `hostNetwork`, the worker Pods have virtual IPs like 10.32.0.5 and this IP will change every time the worker Pod is re-created. If using `hostNetwork`, the worker Pod IP will be the same as the physical Node (using the same network adapter as the machine).

In Spark on K8s, the scheduler will schedule the executor to where the block currently is (a list of block locations returned from `alluxio.hadoop.AbstractFileSystem#getFileBlockLocations`). If the workers are using `hostNetwork`, `getFileBlockLocations` will return the physical Node IPs and the executors will be scheduled there to achieve data locality. Otherwise, Spark schedule will not understand the virtual IPs and fail to find the local Alluxio worker. This means without `hostNetwork`, we cannot have data locality on K8s.

However, `hostNetwork` is usually not permitted in corporation clusters. We have to figure out a way to achieve data locality without `hostNetwork`.

**Describe the solution you'd like**
1. Start the worker Pod without `hostNetwork`. It will have a virtual IP like 10.32.0.5.
1. On the worker side, have properties `ALLUXIO_WORKER_NODE_HOSTNAME` and `ALLUXIO_WORKER_NODE_IP` to feed the worker its node hostname and IP. These properties can be provided with K8s downward API on Pod startup.
1. The worker registers with the master providing these information, in `BlockMasterClient.register`. And the mapping is maintained by the `BlockMaster`.
1. The master, in `alluxio.hadoop.AbstractFileSystem#getFileBlockLocations`, translate the virtual worker addresses to the physical Node hostnames/IPs.

**Describe alternatives you've considered**
As how this is achieve in HDFS on K8s, we update the Spark logic to recognize local HDFS datanode Pods. https://databricks.com/session/hdfs-on-kubernetes-lessons-learned

The downside of this is we now need to wait on the Spark community to merge this change. Also this is cohesive to Spark, not general to all data applications on K8s.

**Urgency**
Medium

**Additional context**

**Comment by jiacheliu3** [comment]
@madanadit Do you think this solution will work?

**Comment by jiacheliu3** [comment]
@madanadit Update to the solution. Now the workers can start without `hostNetwork` but need to use `nodePort` This is verified to work (read/write blocks succeed) in K8s. The illustration PR is https://github.com/Alluxio/alluxio/pull/11194

**The NodePort solution**
1. Start the worker Pod without hostNetwork. It will have a virtual IP like 10.32.0.5.
2. On the worker side, have properties ALLUXIO_WORKER_NODE_HOSTNAME to feed the worker its node hostname. These properties can be provided with K8s downward API on Pod startup. The value of `ALLUXIO_WORKER_NODE_HOSTNAME` will be read into `alluxio.worker.node.hostname`.
3. All the worker properties will be sent to master on register (this is already the case). The master just needs to read the property `alluxio.worker.node.hostname`, and record the worker `virtualHostname : nodeHostname` mapping. Now the master has the mapping.
4. The client side, on startup, sync with the master to get a list of workers. `DefaultBlockMaster#constructWorkerInfoList` answers the call and provides a `List<WorkerInfo>`. Now the master updates the WorkerInfo from virtual hostname to the node hostname.
5. The client will create the read/write streams in `AlluxioBlockStore#getInStream` and `#getOutStream` using the `nodeHostname : nodePort` pair.

The downside is, this solution still needs `nodePort`, which is still something users may have restriction on. The upside is, `nodePort` is much less sensitive than `hostNetwork`.

If `nodePort` is not an option even, the only option is to create a Service for each worker. However this becomes a chicken-egg problem. We need to find a way to feed the worker/master the mapping after the workers are up and services are created.

**Comment by jiacheliu3** [comment]
**No-hostPort solution**

If `hostPort` is not an option, we have to create one service per worker to make it accessible. 

1. Start the worker Pod without hostNetwork. It will have a virtual IP like 10.32.0.5.
2. On the worker side, have properties ALLUXIO_WORKER_NODE_HOSTNAME to feed the worker its node hostname. These properties can be provided with K8s downward API on Pod startup. The value of ALLUXIO_WORKER_NODE_HOSTNAME will be read into alluxio.worker.node.hostname.
3. All the worker properties will be sent to master on register (this is already the case). The master just needs to read the property alluxio.worker.node.hostname, and record the worker virtualHostname : nodeHostname mapping. Now the master has the mapping.
4. The client side, on startup, sync with the master to get a list of workers. DefaultBlockMaster#constructWorkerInfoList answers the call and provides a List<WorkerInfo>. Now the master updates the WorkerInfo from virtual hostname to the node hostname.
5. The client identifies the service for the worker. Now I'm using one service per node (this service only takes the daemon pod on this node). And the service name is has the heuristics with `alluxio-worker-<node IP>`.
6. Now The client will create the read/write streams in `UnderFileSystemFileOutStream#create`, `AlluxioBlockStore#getInStream` and `#getOutStream` using the `<worker-service-name> : hostPort` pair. 

In the current implementation, the service is created manually. The service name cannot be generated manually because we can't know the node IP at run time (The downward API doesn't help on this). The selector also has to be tagged manually on the worker pods.

```
apiVersion: v1
kind: Service
metadata:
  name: alluxio-worker-172-31-92-20
  labels:
    app: alluxio
    chart: alluxio-0.5.5
    release: alluxio
    heritage: Tiller
    role: alluxio-worker
spec:
  ports:
    - port: 29999
      name: rpc
    - port: 30000
      name: web
    - port: 30001
      name: job-rpc
    - port: 30002
      name: job-data
    - port: 30003
      name: job-web
  type: NodePort
  selector:
    role: alluxio-worker
    app: alluxio
    release: alluxio
    kubernetes.io/hostname: 172.31.92.20
```

In the current solution

**Comment by jiacheliu3** [comment]
This has been addressed by #11194
